src/snowflake/snowpark/functions.py

```diff
@@ -3030,18 +3030,27 @@ def substring(
 
     :func:`substr` is an alias of :func:`substring`.
 
-    Example::
+    Example 1::
         >>> df = session.create_dataframe(
         ...     ["abc", "def"],
         ...     schema=["S"],
-        ... ).select(substring(col("S"), 1, 1))
-        >>> df.collect()
+        ... )
+        >>> df.select(substring(col("S"), 1, 1)).collect()
         [Row(SUBSTRING("S", 1, 1)='a'), Row(SUBSTRING("S", 1, 1)='d')]
+
+    Example 2::
+        >>> df = session.create_dataframe(
+        ...     ["abc", "def"],
+        ...     schema=["S"],
+        ... )
+        >>> df.select(substring(col("S"), 2)).collect()
+        [Row(SUBSTRING("S", 2)='bc'), Row(SUBSTRING("S", 2)='ef')]
     """
     s = _to_col_if_str(str, "substring")
     p = pos if isinstance(pos, Column) else lit(pos, _emit_ast=_emit_ast)
-    length = len if isinstance(len, Column) else lit(len, _emit_ast=_emit_ast)
-    return builtin("substring", _emit_ast=_emit_ast)(s, p, length)
+    if len is None:
+        return builtin("substring", _emit_ast=_emit_ast)(s, p)
+    return builtin("substring", _emit_ast=_emit_ast)(s, p, Column._to_expr(len))
```

Comment:
The `Column._to_expr` may cause issues when we add the AST here. We use instead the following and make it explicit:

```
len if isinstance(len, Column) else lit(len)
```

This should also have the correct type then, i.e. not using Expression objects to be passed to the callable created by builtin. Can we change to that?

Also, we're happy to help with adding the AST here. Can we file a ticket onto https://snowflakecomputing.atlassian.net/browse/SNOW-1489960 and add TODO in code with the ticket number to track the work?

If you'd like to give the AST directly a shot, happy to help out and walk you through the process - else also fine with JIRA ticket + TODO and our team will take care of it. Whichever works for you!

Reply:
This is the same usage pattern as `to_date`. Shouldn't that also create problems?

Reply:
We fixed a lot of the `Column._to_expr(...)` cases. There may be some that fell through, merging long-standing feature branches is hard. But yes, we should not use `Column._to_expr(...)` anywhere. Better to be explicit. Happy to discuss/share more in detail.